pkg/slices/slices.go

```diff
@@ -0,0 +1,106 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package slices defines various functions useful with slices of any type.
+package slices
+
+import (
+	"golang.org/x/exp/constraints"
+	"golang.org/x/exp/slices"
+)
+
+// Equal reports whether two slices are equal: the same length and all
+// elements equal. If the lengths are different, Equal returns false.
+// Otherwise, the elements are compared in increasing index order, and the
+// comparison stops at the first unequal pair.
+// Floating point NaNs are not considered equal.
+func Equal[E comparable](s1, s2 []E) bool {
+	return slices.Equal(s1, s2)
+}
+
+// SortFunc sorts the slice x in ascending order as determined by the less function.
+// This sort is not guaranteed to be stable.
+// The slice is modified in place but returned.
+func SortFunc[E any](x []E, less func(a, b E) bool) []E {
+	slices.SortFunc(x, less)
+	return x
+}
+
+// Sort sorts a slice of any ordered type in ascending order.
+// The slice is modified in place but returned.
+func Sort[E constraints.Ordered](x []E) []E {
+	slices.Sort(x)
+	return x
+}
+
+// Clone returns a copy of the slice.
+// The elements are copied using assignment, so this is a shallow clone.
+func Clone[S ~[]E, E any](s S) S {
+	return slices.Clone(s)
+}
+
+// Delete removes the element i from s, returning the modified slice.
+func Delete[S ~[]E, E any](s S, i int) S {
+	return slices.Delete(s, i, i+1)
+}
+
+// Contains reports whether v is present in s.
+func Contains[E comparable](s []E, v E) bool {
+	return slices.Contains(s, v)
+}
+
+// FindFunc finds the first element matching the function, or nil if none do
+func FindFunc[E any](s []E, f func(E) bool) *E {
```

Comment:
What if it is a []*pointer, but with an elment  == nil.

Reply:
In that case, `E=*pointer`. So `**E` is the response.

If we made it `func FindFunc[E any](s []E, f func(E) bool) E` then you would not be able to distinguish in your case.

We could make it `func FindFunc[E any](s []E, f func(E) bool) (E, bool)` but I think its easier to use as written here